CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt

```diff
@@ -69,22 +68,27 @@ class CardClient internal constructor(
                 is ConfirmPaymentSourceResult.Success -> {
                     if (response.payerActionHref == null) {
                         analytics.notifyApproveOrderSucceeded(response.orderId)
-                        val result =
-                            response.run { CardResult(orderId = orderId, status = status?.name) }
-                        approveOrderListener?.onApproveOrderSuccess(result)
+                        val result = response.run {
+                            CardApproveOrderResult.Success(
+                                orderId = orderId,
+                                status = status?.name
+                            )
+                        }
+                        callback.onCardApproveOrderResult(result)
                     } else {
                         analytics.notifyApproveOrderAuthChallengeReceived(cardRequest.orderId)
-                        approveOrderListener?.onApproveOrderThreeDSecureWillLaunch()
 
                         val url = Uri.parse(response.payerActionHref)
                         val authChallenge = CardAuthChallenge.ApproveOrder(url, cardRequest)
-                        approveOrderListener?.onApproveOrderAuthorizationRequired(authChallenge)
+                        val result = CardApproveOrderResult.AuthorizationRequired(authChallenge)
```

Comment:
It's interesting that you have this intermediate state on the same result type.

Reply:
This method does have three possible results. Alternatively, I thought about bundling the `payer_action_required` status into the `Failure` case but this felt more natural.

Technically it isn't a failure. PPaaS just needs more information before the buyer can continue.